WordPress/gutenberg
# Tabs: Prevent accidental overflow in indicator

## What?

This PR takes the decimal point into account when calculating the indicator's style (top, left, width, height) to prevent the indicator from overflowing.

## Why?

My understanding is that `offset{top|left|width|height}` rounds the decimal point. This error may cause the indicator to overflow.

One problem I've seen is that inspector controls overflow on Windows OS.

![inspector-controls-overflow](https://github.com/WordPress/gutenberg/assets/54422211/5e7f1078-bf49-422c-bdbb-b9f20a20f65d)

The reasons why this occurs are as follows.

- The sidebar including inspector controls is [`280px` wide](https://github.com/WordPress/gutenberg/blob/62b0efd7af66c18d5717028e75cf8ff0c441c196/packages/interface/src/components/complementary-area/index.js#L45).
- If the inspector control's content height is higher than the box height, a vertical scrollbar will appear.
- In my Chrome, the scrollbar width is `17px`. As a result, the width of the inspector control is `263px`.
  ![image](https://github.com/WordPress/gutenberg/assets/54422211/fb3e9e07-d19b-48b4-b877-f998a3ca14eb)
- Tab button width is `131.5px`.
  ![image](https://github.com/WordPress/gutenberg/assets/54422211/68796eef-044e-4111-a543-76b1ea258417)
- The current logic rounds the decimal point in the indicator's style calculation, so when the button on the right is selected, the CSS variable has a value like this:
  ![image](https://github.com/WordPress/gutenberg/assets/54422211/fac4d1e1-2038-4196-b14b-e2175ad5dad9)
- The left and top values should be `131.5px`, but they are actually `132px`, so the indicator overflows by 1px.

## How?

~~This could easily be solved by applying `overflow:hidden` to the [TabListWrapper](https://github.com/WordPress/gutenberg/blob/62b0efd7af66c18d5717028e75cf8ff0c441c196/packages/components/src/tabs/styles.ts#L14), but I chose to use `getBoundingClientRect()` to generate an accurate indicator style that takes decimal points into account.~~

In response to the feedback, a different approach was adopted - see [the comments starting here](https://github.com/WordPress/gutenberg/pull/61979#issuecomment-2131437104) for more details.

## Testing Instructions

On MacOS, scrollbars have no physical width, so you won't be able to reproduce this problem by default. However, you should be able to reproduce the horizontal scroll bar issue on MacOS by following the steps below.

- Set the value [here](https://github.com/WordPress/gutenberg/blob/62b0efd7af66c18d5717028e75cf8ff0c441c196/packages/interface/src/components/complementary-area/index.js#L45) to `279`
- Change OS settings to always display scroll bars
- Select any block and open the block sidebar
- Select the "Style" tab
   - trunk: CSS variable has the following value and causes an overflow.
    ![image](https://github.com/WordPress/gutenberg/assets/54422211/9c9eba53-aee8-4c26-8b34-e3969a865fbd)
  - This PR: CSS variables contain decimal values and overflow should not occur
    ![image](https://github.com/WordPress/gutenberg/assets/54422211/4a836982-07aa-4ae0-8242-d571d0602cd0)
